### PR TITLE
[P2P] Remove RelatedPeerConnector

### PR DESCRIPTION
**Pull request by fassadlr**
These changes include:

- Remove RelatedPeerConnector
- Remove WellKnownPeerConnectorSelectors
- Inject IConnectionManager into PeerConnector
- Add Integration test for PeerConnectorAddNode to check that an already connected node cannot be reselected.

### CODE REVIEW

File: src/Stratis.Bitcoin.IntegrationTests/P2P/NodeConnectionTests.cs [hunk 1]
```diff
@@ -1,98 +0,0 @@
-﻿using NBitcoin;
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.IntegrationTests/P2P/NodeConnectionTests.cs]: FYI. This file got renamed and a second test added.

File: src/Stratis.Bitcoin/P2P/PeerConnector.cs [hunk 2]
```diff
@@ -142,21 +131,25 @@ public abstract class PeerConnector : IPeerConnector
...
 
         /// <summary>Constructor used for unit testing.</summary>
-        protected PeerConnector(NodeSettings nodeSettings, IPeerAddressManager peerAddressManager)
+        protected PeerConnector(ILoggerFactory loggerFactory, NodeSettings nodeSettings, IPeerAddressManager peerAddressManager)
         {
             Guard.NotNull(nodeSettings, nameof(nodeSettings));
             Guard.NotNull(peerAddressManager, nameof(peerAddressManager));
 
+            this.connectedPeers = new NetworkPeerCollection();
+            this.loggerFactory = loggerFactory;
```
Comment (dangershony) [inline src/Stratis.Bitcoin/P2P/PeerConnector.cs]: `NodeLifetime` must be injected

File: src/Stratis.Bitcoin/P2P/PeerConnector.cs [hunk 3]
```diff
@@ -142,21 +131,25 @@ public abstract class PeerConnector : IPeerConnector
         }
 
         /// <summary>Constructor used for unit testing.</summary>
-        protected PeerConnector(NodeSettings nodeSettings, IPeerAddressManager peerAddressManager)
+        protected PeerConnector(ILoggerFactory loggerFactory, NodeSettings nodeSettings, IPeerAddressManager peerAddressManager)
         {
             Guard.NotNull(nodeSettings, nameof(nodeSettings));
             Guard.NotNull(peerAddressManager, nameof(peerAddressManager));
 
+            this.connectedPeers = new NetworkPeerCollection();
```
Comment (dangershony) [inline src/Stratis.Bitcoin/P2P/PeerConnector.cs]: I guess we can keep this as null as its set in `Initialize` does it make sense to put an instance here?

File: src/Stratis.Bitcoin/P2P/PeerConnector.cs [hunk 4]
```diff
@@ -188,14 +181,10 @@ public void RemovePeer(NetworkPeer peer)
...
         /// </summary>
-        /// <remarks>
-        /// TODO: This will be removed when we remove related peer connectors.
-        /// </remarks>
         /// <param name="ipEndpoint">The endpoint to check.</param>
         internal bool IsPeerConnected(IPEndPoint ipEndpoint)
         {
-            bool peerIsConnected = this.RelatedPeerConnector.GlobalConnectedNodes().Any(a => this.GroupSelector(a).SequenceEqual(this.GroupSelector(ipEndpoint)));
-            return peerIsConnected;
+            return this.connectedPeers.FindByEndpoint(ipEndpoint) != null;
```
Comment (dangershony) [inline src/Stratis.Bitcoin/P2P/PeerConnector.cs]: So much easier

File: src/Stratis.Bitcoin/P2P/PeerDiscoveryLoop.cs [hunk 5]
```diff
@@ -82,16 +82,18 @@ public sealed class PeerDiscovery : IPeerDiscovery
...
         /// <inheritdoc/>
-        public void DiscoverPeers(NetworkPeerConnectionParameters parameters)
+        public void DiscoverPeers(IConnectionManager connectionManager)
         {
             // If peers are specified in the -connect arg then discovery does not happen.
             if (this.nodeSettings.ConnectionManager.Connect.Any())
                 return;
 
-            if (!parameters.PeerAddressManagerBehaviour().Mode.HasFlag(PeerAddressManagerBehaviourMode.Discover))
+            if (!connectionManager.Parameters.PeerAddressManagerBehaviour().Mode.HasFlag(PeerAddressManagerBehaviourMode.Discover))
```
Comment (dangershony) [inline src/Stratis.Bitcoin/P2P/PeerDiscoveryLoop.cs]: Doesn't the previous line obsolete this line?

File: src/Stratis.Bitcoin/P2P/PeerConnector.cs (hunk 3 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerConnector.cs]: @dangershony I've removed this constructor in PR #891 👍

File: src/Stratis.Bitcoin/P2P/PeerConnector.cs (hunk 2 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerConnector.cs]: Same here, these "unit test" constructors has been dropped in #891

File: src/Stratis.Bitcoin/P2P/PeerConnector.cs (hunk 4 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerConnector.cs]: 💯

File: src/Stratis.Bitcoin/P2P/PeerDiscoveryLoop.cs (hunk 5 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerDiscoveryLoop.cs]: How do you mean @dangershony ? First we check to see of there are Connect nodes specified at startup and then, if not, we check to see if ```PeerAddressManagerBehaviour``` is set to ```Discover```?

I suppose we can combine that if statement?

WDYT

File: src/Stratis.Bitcoin.IntegrationTests/P2P/PeerConnectionTests.cs [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+                this.parameters,
+                this.peerAddressManager,
+                new IPeerConnector[] { peerConnectorDiscovery },
+                peerDiscovery);
+
+            // Start peer discovery.
+            peerDiscovery.DiscoverPeers(connectionManager);
+
+            // Wait until we have discovered 10 peers.
+            TestHelper.WaitLoop(() => this.peerAddressManager.Peers.Count > 10);
```
[the remark this answers is not in the document]
Comment (fassadlr) [reply src/Stratis.Bitcoin.IntegrationTests/P2P/PeerConnectionTests.cs]: I just thought it'd be good to have an integration test actually tries to discover peers and then connect to them but I understand your argument with regards to loads of people running them. Should I possibly then remove this test?

File: src/Stratis.Bitcoin/P2P/PeerDiscoveryLoop.cs (hunk 5 above)
Comment (dangershony) [reply src/Stratis.Bitcoin/P2P/PeerDiscoveryLoop.cs]: No I mean if there are `connect` nodes then the behaviour will never be set to `discover` right?

File: src/Stratis.Bitcoin.IntegrationTests/P2P/PeerConnectionTests.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (dangershony) [reply src/Stratis.Bitcoin.IntegrationTests/P2P/PeerConnectionTests.cs]: If you can find a way to mock the processes without connecting externally then good else not delete it.

[the remark this answers is not in the document]
Comment (dangershony) [reply src/Stratis.Bitcoin.IntegrationTests/P2P/PeerConnectionTests.cs]: Maybe @Neurosploit has an idea on how to test this.

File: src/Stratis.Bitcoin/P2P/PeerDiscoveryLoop.cs (hunk 5 above)
Comment (dangershony) [reply src/Stratis.Bitcoin/P2P/PeerDiscoveryLoop.cs]: Ah the `PeerAddressManagerBehaviour` is global so the flag is set once ok thats fine

**Pull request merged**